text_dir_codepoint lint update

enthropy7 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/cargo/pull/16452)*

### What does this PR try to resolve?

This PR implements the **text_direction_codepoint lint** for `Cargo.toml` files:
Fixes #16373 and #16374

_lint detects Unicode BiDi (bidirectional) control codepoints that can be used in "Trojan Source" attacks. These invisible characters can alter the visual display of text without changing its underlying representation, potentially making malicious manifest content appear benign. All like descritped in rustc's text_direction_codepoint_in_comment/literal lint but for Cargo manifests._

Detected codepoints:

U+202A LEFT-TO-RIGHT EMBEDDING
U+202B RIGHT-TO-LEFT EMBEDDING
U+202C POP DIRECTIONAL FORMATTING
U+202D LEFT-TO-RIGHT OVERRIDE
U+202E RIGHT-TO-LEFT OVERRIDE
U+2066 LEFT-TO-RIGHT ISOLATE
U+2067 RIGHT-TO-LEFT ISOLATE
U+2068 FIRST STRONG ISOLATE
U+2069 POP DIRECTIONAL ISOLATE

Default level: `deny`

### How to test and review this PR?

`cargo test -p cargo --test testsuite -- lints::text_direction_codepoint`

### Manual testing:

Create a manifest with BiDi codepoint (U+202E)

```
printf '[package]\nname = "foo"\nversion = "0.1.0"\nedition = "2024"\ndescription = "A \xE2\x80\xAEtest"\n' > Cargo.toml
mkdir -p src && echo 'fn main(){}' > src/main.rs
cargo +nightly check -Zcargo-lints
```

The lint can be configured via `[lints.cargo]`:
```
[lints.cargo]
text_direction_codepoint = "allow"  # or "warn"
```

epage commented:
Feel free to clean up the commits for how they should be reviewed and merged, rather than reflecting the implementation.

On top of that, our [contrib guide](https://doc.crates.io/contrib/process/working-on-cargo.html#submitting-a-pull-request) suggests adding tests in a commit before the feature, showing the current behavior.  The feature commit would then update the tests to show the new behavior.  When reviewing the feature commit, this helps remove all of the boilerplate from the view and see exactly what changed in behavior.

epage commented on src/cargo/core/workspace.rs:
> Fixes https://github.com/rust-lang/cargo/issues/16373 and https://github.com/rust-lang/cargo/issues/16374

I had forgotten that we had separate issues.  The question is whether we should mirror rustc or lint them together.  I had overlooked this conversation in #16374, focusing on the other parts.

weihanglo replied:
This sounds reasonable to me. Thanks for the investigation!

Commit:
Add tests for text_direction_codepoint lint

Tests for detecting Unicode BiDi control codepoints in Cargo.toml.
Currently these pass as there is no lint implementation yet.

enthropy7 commented:
> > I can restructure commits if needed, but the tests wouldn't show meaningful info because before there was no lint to trigger. sorry btw, for next PR's i would take that info
> 
> "before there was no lint to trigger": that is exactly what we want to see with this.

Done, restructured commits

enthropy7 commented:
if anything is need - i’m here 24/7 :3 just wanna merge it and take the next task

weihanglo commented on src/cargo/core/workspace.rs:
```diff
@@ -1370,6 +1378,17 @@ impl<'gctx> Workspace<'gctx> {
             )?;
+            // Only check for BiDi codepoints in virtual workspace manifests.
+            // For package roots, the lint is run in emit_pkg_lints.
+            if matches!(self.root_maybe(), MaybePackage::Virtual(_)) {
```
Should we move this logic into the lint function itself?


In the future we might want to make the lint infra a `LintContext`, and register lints all at once, instead of the current manual calling functions.

enthropy7 replied:
corrected it, thanks, i haven't thought about it (my bad)

weihanglo commented on tests/testsuite/lints/text_direction_codepoint.rs:
```diff
@@ -0,0 +1,242 @@
+  |                       ^ this invisible unicode codepoint changes text flow direction
+  |
+  = [HELP] if their presence wasn't intentional, you can remove them
+[ERROR] encountered 2 errors while running lints
```
(Not a blocker)

Do we want to have one error per finding, or one per line, or whathever rustc does if making sense to cargo? I found rustc's approach looks nicer btw

```
error: unicode codepoint changing visible direction of text present in comment
  --> $DIR/unicode-control-codepoints.rs:40:1
   |
LL | //"/*� } �if isAdmin� � begin admins only */"
   | ^^^^^-^^^-^^^^^^^^^^-^-^^^^^^^^^^^^^^^^^^^^^^
   | |    |   |          | |
   | |    |   |          | '\u{2066}'
   | |    |   |          '\u{2069}'
   | |    |   '\u{2066}'
   | |    '\u{202e}'
   | this comment contains invisible unicode text flow control codepoints
   |
   = note: these kind of unicode codepoints change the way text flows on applications that support them, but can cause confusion because they change the order of characters on the screen
   = help: if their presence wasn't intentional, you can remove them

```

https://github.com/rust-lang/rust/blob/fcd630976c460c819c4bbcaf107d0c94501205d8/tests/ui/parser/unicode-control-codepoints.stderr#L212-L225

weihanglo reviewed (commented):
You might need to fix CI job failures (rustfmt).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/16452/5d927cd53cbb3548ded28f35bb2b2b051172568b..32eb70985842acd85e815466e1aa1569d897903b)*

enthropy7 commented:
done, i think , thanks for all the help

weihanglo reviewed (commented):
Other than the question, looks good to me.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/16452/d27e95e48583613a0165cfa66eaaea7644cf7397..b20cc7000eb0ca0e08169507acf4db8f01984d2e)*

Commit:
Implement text_direction_codepoint lint

epage commented:
As resolving this is one of the steps towards stabilizing #16950, I've gone ahead and taken a crack at this, applying what review feedback I saw and applying what I would have said.

See #16950.
